Shader preprocessor: fix macro expansion inconsistencies with glsl

pirey0 opened the pull request:
fixes Issues 1,2,3,5,6,8 from #96253 

- fixed shader preprocessor macro expansion order
    - Previously: Macros where expanded in definition order: this caused inconsistencies with glsl and could result in different generated code when moving defines around ⚠️ 
    - Now: Marcos are expanded left-to-right as they appear in the code with marco-arguments being expanded before the body to match glsl more closely
- adjusted macro expansion recursion to match glsl
    - Previously: Macro recursion was allowed up to an arbitrary depth ⚠️ 
    - Now: Like in glsl a macro connot be expanded again within it's own expansion
- added shader preprocessor support for 0 argument funktionline macros
    - Previously: functionlike macros with 0 arguments (`#define macro()`) threw an  error 
    - Now: work like in glsl
- fixed functionlike macro argument expansion re-expanding already replaced macros
    - Previously: subsequent arguments in a macro could expand sections from previous arguments (see issue 8)

- refactored macro expansion
- added multiple tests

pirey0 committed:
fixed shader preprocessor macro expansion inconsistencies with glsl

fixed shader preprocessor macro expansion order and recursion
added shader preprocessor support for 0 argument funktionline macros
fixed shader functionlike macros expansion with 0 arguments
fixed functionlike macro argument expansion re-expanding already replaced macros
refactored macro expansion
added multiple tests

geekley commented:
Thanks for the quick PR!
Might be prudent to also add a test case for the opposite just in case, e.g.:
```glsl
#define nonfunctionlike 42
const int a = nonfunctionlike + nonfunctionlike();
// expects: const int a = 42 + 42();
// no preprocessor error
```

pirey0 commented:
I've updated the PR to also include fixes for issues 5,6 and 8 from #96253
( See no reason to open a second PR since the two would have been tightly connected and this hasn't been reviewed yet.)

If this gets merged we can close #96253 since issue 4 is superfluous. (discussed in the 2024-09-03 rendering meeting.)

I also added more preprocessor tests using [glsl](https://github.com/KhronosGroup/glslang) as a reference point.
